Added events to release and evolve pokemon

andythorne opened the pull request:
### Short Description:

While the bot does a great job at automating pokemon actions, I found myself wanting to do specific actions manually. This change adds support for event listening from web client requests to evolve and 
release pokemon. The events can be enabled with the `web_events` option (`True` for all events, or and array of event actions it enable (`['release_pokemon', 'evolve_pokemon']`). Default is `False` to disable all.

There _could_ be a problem if someone decides to put this on a public webserver as it will let anyone evolve and release your pokemon, which is why they are all off be default.

Frontend implementation here: https://github.com/OpenPoGo/OpenPoGoWeb/pull/134. If that is merged first i'll update this PR with the new subproject sha.
### Changes:
- Added `bot_initialized` event to fire after bot has been setup
- Added `web_events` plugin to listen to incoming events form the web client
- Added `config.web_events` key to configure what events are enabled. `True` is all events, `False` is none, or provide an array of event names to enable. Default is `False`.

@OpenPoGo/maintainers

andythorne committed:
Added web events to release and evolve pokemon

MaxLeiter commented:
Awesome!
:+1: 

Would appreciate both @wchill and @Reaver01 :+1:'ing this PR

wchill commented:
This will need rebasing, but looks fine to me otherwise.

andythorne commented:
Rebased :+1:

MaxLeiter commented:
@Reaver01 waiting on you :)

wchill commented on pokemongo_bot/utils.py:
```diff
@@ -145,3 +145,5 @@ def convert_to_utf8(data):
     elif isinstance(data, set):
         return set(map(convert_to_utf8, data))
+    else:
+        return str(data)
```
Why do we need this?

andythorne replied:
There was an issue when the utf8 encoding the json where the pokemon unique ID was being stripped out as it didn't match in this method.

jadbalout commented:
Why not work on dyanmic conversation if the client is already running? In other words, we need to start working on the Socket IO even more. I'll make a PR that has socket conversations about evolving and releasing pokemons.

wchill commented:
Okay, now that I actually read what this is doing, it seems that you register event handlers for transferring and evolving Pokemon, then fire the events from the web client?

andythorne commented:
Yep that's correct, so you can click a button on the UI and it will transfer or evolve a pokemon.

I'll refactor this tomorrow for the new EM.

andythorne commented:
Ok, i've been looking at moving this to a module, but I've come up against some blockers:

Both events need the `bot` to run their actions. This is usually done with `manager.fire_with_context(...)`, but the event is fired will be fired by a module, which has no knowledge of the bot. To get this working it seems we either:

- Inject the `bot` into the `event_manager`, or
- Inject the `bot` into the module

Any suggestions? Python isn't my 1st language so I not sure if there is a way to do it I don't know about!

If we can't do that I can't see how the actions will be able to work.

wchill commented:
I thought about situations like this. In this kind of situation, what we can do is have the bot fire a `bot_initialized` event or something, passing itself as the context. Any modules that require a long-lived copy of the bot context like this module must keep track of it themselves by hooking the `bot_initialized` event and saving it to some data structure.

For example:

```
from pokemongo_bot.event_manager import manager

bot_contexts = dict()


@manager.on("bot_initialized")
def bot_initialized_event(bot):
    bot_contexts[bot.username] = bot


def some_function_that_needs_bot_context(username):
    bot_contexts[username].fire("event_requiring_bot_context")
```

This example code should work for multiple bot instances as well, as long as they are all using the same event manager. (Currently this means that all bots must run in the same process, though we haven't technically added support for multi user yet.)

andythorne commented:
Sounds like a good solution. I'll make a seperate PR to implement a `bot_initialized` event.

andythorne commented:
Updated and rebased. This is now a plugin!

@wchill @MaxLeiter

MaxLeiter commented:
:+1: woot!

wchill commented on plugins/web_events/__init__.py:
```diff
@@ -0,0 +1,70 @@
+                log('[#] Releasing pokemon with id {}'.format(pokemon_name), 'green')
+
+                bot.api_wrapper.release_pokemon(pokemon_id=int(payload['pokemon_id']))
+                bot.api_wrapper.call()
```
Instead of calling the API directly, this should just fire `transfer_pokemon` with parameter `transfer_list=[Pokemon to transfer]`.

wchill commented on plugins/web_events/__init__.py:
```diff
@@ -0,0 +1,70 @@
+                pokemon_name = bot.pokemon_list[pokemon.pokemon_id - 1]["Name"]
+                log('[#] Evolving {}'.format(pokemon_name), 'green')
+
+                bot.api_wrapper.release_pokemon(pokemon_id=int(payload['pokemon_id']))
```
This isn't evolving, just releasing?

andythorne replied:
:dizzy_face:  copy/paste fail. i'll get that fixed.

andythorne commented:
Updated.

I added an `evolve_pokemon` module and prefixed all the `web_events` module's events with `web_` to there won't be any conflicts or confusion on which event to fire.

andythorne commented:
Done in https://github.com/OpenPoGo/OpenPoGoBot/pull/136

MaxLeiter commented:
@andythorne please ping maintainers when you feel this is ready for review

andythorne committed:
removed evolve_pokemon

andythorne commented:
I'm going to close this due to #244
